src/com/reco1l/osu/ui/Preference.kt

```diff
@@ -0,0 +1,308 @@
+                Option(entries[it].toString(), values[it])
+            })
+        }
+
+        dialog.setTitle(title.toString())
+        dialog.setOnSelectListener { value = it as String }
+    }
+
+
+    override fun onClick() = dialog.show()
+
+
+    override fun onGetDefaultValue(a: TypedArray, index: Int) = a.getString(index)
+
+    override fun onSetInitialValue(defaultValue: Any?) {
+        value = getPersistedString(defaultValue as String?)
+    }
+
+
+    override fun onSaveInstanceState(): Parcelable? {
+        val superState = super.onSaveInstanceState()
+        return if (isPersistent) superState else SavedState(superState).also { it.value = value }
+    }
+
+    override fun onRestoreInstanceState(state: Parcelable?) {
+
+        if (state == null || state !is SavedState) {
+            super.onRestoreInstanceState(state)
+            return
+        }
+
+        super.onRestoreInstanceState(state.superState)
+        value = state.value
+    }
+
+
+    private class SavedState : BaseSavedState {
+
+        constructor(superState: Parcelable?) : super(superState)
+
+        constructor(source: Parcel) : super(source) {
+            value = source.readString()
+        }
+
+        var value: String? = null
+
+        override fun writeToParcel(dest: Parcel, flags: Int) {
+            super.writeToParcel(dest, flags)
+            dest.writeString(value)
+        }
+
+    }
+}
+
+
+/// Input
+
+class InputView(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : androidx.appcompat.widget.AppCompatEditText(context, attrs, defStyleAttr) {
+
+    constructor(context: Context): this(context, null)
+    constructor(context: Context, attrs: AttributeSet?): this(context, attrs, android.R.attr.editTextStyle)
+
+    override fun onKeyPreIme(keyCode: Int, event: KeyEvent): Boolean {
+        if (keyCode == KeyEvent.KEYCODE_BACK && event.action == KeyEvent.ACTION_UP) {
+            hideKeyboard()
+        }
+        return super.onKeyPreIme(keyCode, event)
+    }
+}
+
+
+@SuppressLint("RestrictedApi")
+class InputPreference(context: Context, attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int): Preference(context, attrs, defStyleAttr, defStyleRes) {
+
+    constructor(context: Context) : this(context, null)
+    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : this(context, attrs, defStyleAttr, 0)
+    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, TypedArrayUtils.getAttr(context, R.attr.editTextPreferenceStyle, android.R.attr.editTextPreferenceStyle))
+
+
+    init {
+        val attributes = context.obtainStyledAttributes(attrs, R.styleable.Preference, defStyleAttr, defStyleRes)
+
+        val layout = TypedArrayUtils.getResourceId(
+            attributes,
+            R.styleable.Preference_layout,
+            R.styleable.Preference_android_layout,
+            ru.nsu.ccfit.zuev.osuplus.R.layout.settings_preference_input
+        )
```

Comment:
This is equivalent to the following code

```suggestion
        val layout = attributes.getResourceId(R.styleable.Preference_layout, attributes.getResourceId(R.styleable.Preference_android_layout, ru.nsu.ccfit.zuev.osuplus.R.layout.settings_preference_input))
```